### PR TITLE
[Tune] Pbt Function API

### PR DESCRIPTION
**Pull request by amogkam**
Adds a convnet tuning example with PBT using Function API and updates PBT guide to show this example. Also adds a unit test for PBT function API checkpointing. Includes @krfricke's fix for function API checkpointing.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/tune/examples/pbt_convnet_function_example.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+    analysis = tune.run(
+        train_convnet,
+        name="pbt_test",
+        scheduler=scheduler,
+        verbose=1,
+        stop=stopper,
+        export_formats=[ExportFormat.MODEL],
+        checkpoint_score_attr="mean_accuracy",
+        checkpoint_freq=5,
```
Comment (richardliaw) [inline python/ray/tune/examples/pbt_convnet_function_example.py:109]: I don't think it makes sense to have checkpoint_freq with the Function API

Comment (amogkam) [reply python/ray/tune/examples/pbt_convnet_function_example.py:109]: Yeah I think it just gets ignored. I'll remove it from here and add a note to the docs. But it shouldn't be causing an error right?

Comment (richardliaw) [reply python/ray/tune/examples/pbt_convnet_function_example.py:109]: I don't have an exact explanation of the error - the fact that there's an error means that my mental model of how checkpointing works is broken. 

All i know is that this configuration has close proximity to the error (conceptually speaking), so I would try turning it off.

File: python/ray/tune/checkpoint_manager.py [hunk 2]
```diff
@@ -137,7 +137,7 @@ def on_checkpoint(self, checkpoint):
                 self._membership.remove(worst)
             # Don't delete the newest checkpoint. It will be deleted on the
             # next on_checkpoint() call since it isn't in self._membership.
-            if worst != checkpoint:
+            if worst.value != checkpoint.value:
```
Comment (richardliaw) [inline python/ray/tune/checkpoint_manager.py:140]: nice catch!

File: python/ray/tune/examples/pbt_convnet_example.py [hunk 3]
```diff
@@ -133,7 +133,7 @@ def stop_all(self):
 
     best_trial = analysis.get_best_trial("mean_accuracy")
     best_checkpoint = max(
-        analysis.get_trial_checkpoints_paths(best_trial, "mean_accuracy"))
+        analysis.get_trial_checkpoints_paths(best_trial, "mean_accuracy"), key=lambda x:x[1])
```
Comment (richardliaw) [inline python/ray/tune/examples/pbt_convnet_example.py:136]: hmm, this is a little gross - does it make sense to provide a helper function in analysis?

File: python/ray/tune/tests/test_trial_scheduler_pbt.py [hunk 4]
```diff
@@ -83,6 +103,34 @@ def testPermutationContinuation(self):
...
+        tune.run(
+            MockTrainingFunc,
+            config={
+                "a": tune.sample_from(lambda _: param_a()),
+                "b": tune.sample_from(lambda _: param_b()),
+                "c": 1
+            },
+            fail_fast=True,
+            num_samples=20,
+            global_checkpoint_period=1,
```
Comment (richardliaw) [inline python/ray/tune/tests/test_trial_scheduler_pbt.py:127]: we don't need this right?

File: python/ray/tune/tests/test_trial_scheduler_pbt.py [hunk 5]
```diff
@@ -32,6 +32,26 @@ def load_checkpoint(self, tmp_checkpoint_dir):
...
+    iter = 0
+    a = config["a"]
+    b = config["b"]
+    # c = config["c"]
+
+    if checkpoint_dir:
+        checkpoint_path = os.path.join(checkpoint_dir, "model.mock")
+        with open(checkpoint_path, "rb") as fp:
+            a, b = pickle.load(fp)
+
```
Comment (richardliaw) [inline python/ray/tune/tests/test_trial_scheduler_pbt.py:45]: do you need to reset "iter"?

File: python/ray/tune/examples/pbt_convnet_function_example.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        path = os.path.join(checkpoint_dir, "checkpoint")
+        checkpoint = torch.load(path)
+        model.load_state_dict(checkpoint["model_state_dict"])
+        step = checkpoint["step"]
+
+    while True:
+        train(model, optimizer, train_loader)
+        acc = test(model, test_loader)
+        if step % 5 == 0:
+            with tune.checkpoint_dir(step=step) as checkpoint_dir:
```
Comment (richardliaw) [inline python/ray/tune/examples/pbt_convnet_function_example.py:42]: can you add comments about what you're doing here in the code?

File: doc/source/tune/_tutorials/tune-advanced-tutorial.rst [hunk 7]
```diff
@@ -145,6 +145,38 @@ thus just use the same ``Trainable`` for the replay run.
...
+
+Now we can kick off the tuning process by invoking tune.run, setting ``resuse_actors`` to ``True``:
+
+.. literalinclude:: /../../python/ray/tune/examples/pbt_convnet_example.py
+   :language: python
+   :start-after: __tune_begin__
+   :end-before: __tune_end__
 
 DCGAN with Trainable and PBT
 ----------------------------
```
Comment (richardliaw) [inline doc/source/tune/_tutorials/tune-advanced-tutorial.rst:182]: does it make sense to also port DCGAN to the class API?

File: doc/source/tune/_tutorials/tune-advanced-tutorial.rst [hunk 8]
```diff
@@ -145,6 +145,38 @@ thus just use the same ``Trainable`` for the replay run.
         scheduler=replay,
         stop={"training_iteration": 100})
 
+Trainable Class API with Population Based Training
+--------------------------------------------------
+
+While using the function API is sufficient for most PBT workflows, sometimes it is better to use the Trainable Class API.
+For example, if your training requires setting up an expensive reinforcement learning environment, we don't want to start a
+new trial on every single exploit. If we use the Trainable Class API, we can resue actors if our trainable implements ``reset_config``
+and we can avoid the expensive setup.
```
Comment (richardliaw) [inline doc/source/tune/_tutorials/tune-advanced-tutorial.rst:154]: IMO I think we can get rid of this section (Trainable Class API)

**Review by richardliaw — COMMENTED** [review]
left a couple comments!

File: python/ray/tune/checkpoint_manager.py (hunk 2 above)
Comment (amogkam) [reply python/ray/tune/checkpoint_manager.py:140]: Thanks to @krfricke!

File: python/ray/tune/examples/pbt_convnet_example.py (hunk 3 above)
Comment (amogkam) [reply python/ray/tune/examples/pbt_convnet_example.py]: I added a new function to do this in Analysis.

File: python/ray/tune/tests/test_trial_scheduler_pbt.py (hunk 4 above)
Comment (amogkam) [reply python/ray/tune/tests/test_trial_scheduler_pbt.py]: removed

File: python/ray/tune/tests/test_trial_scheduler_pbt.py (hunk 5 above)
Comment (amogkam) [reply python/ray/tune/tests/test_trial_scheduler_pbt.py]: I changed the test do this, but I don't think it actually matters for the test.

File: python/ray/tune/examples/pbt_convnet_function_example.py (hunk 6 above)
Comment (amogkam) [reply python/ray/tune/examples/pbt_convnet_function_example.py:47]: Added comments.

File: doc/source/tune/_tutorials/tune-advanced-tutorial.rst (hunk 7 above)
Comment (amogkam) [reply doc/source/tune/_tutorials/tune-advanced-tutorial.rst:149]: I think so. But I'll save that for a later PR.

File: doc/source/tune/_tutorials/tune-advanced-tutorial.rst (hunk 8 above)
Comment (amogkam) [reply doc/source/tune/_tutorials/tune-advanced-tutorial.rst]: Removed

**Review by richardliaw — APPROVED** [review]

**Comment by richardliaw** [comment]
nice! ping when tests pass (lint is failing right now)

**Comment by richardliaw** [comment]
bump tests still failing

File: python/ray/tune/tests/test_trial_scheduler_pbt.py [hunk 9]
```diff
@@ -83,6 +108,33 @@ def testPermutationContinuation(self):
...
+                "b": tune.sample_from(lambda _: param_b()),
+                "c": 1
+            },
+            fail_fast=True,
+            num_samples=20,
+            keep_checkpoints_num=1,
+            checkpoint_score_attr="min-training_iteration",
+            scheduler=scheduler,
+            name="testPermutationContinuationFunc",
+            stop={"training_iteration": 5})
```
Comment (richardliaw) [inline python/ray/tune/tests/test_trial_scheduler_pbt.py:136]:
```suggestion
            num_samples=4,
            keep_checkpoints_num=1,
            checkpoint_score_attr="min-training_iteration",
            scheduler=scheduler,
            name="testPermutationContinuationFunc",
            stop={"training_iteration": 3})
```

**Comment by amogkam** [comment]
@richardliaw tests passing

**Pull request merged**